chainstate/tx-verifier/src/transaction_verifier/mod.rs

```diff
@@ -582,6 +542,87 @@ impl<'a, S: TransactionVerifierStorageRef, U: UtxosView> TransactionVerifier<'a,
         }
     }
 
+    pub fn connect_pos_accounting_outputs(
+        &mut self,
+        tx_source: TransactionSource,
+        tx: &Transaction,
+    ) -> Result<(), ConnectTransactionError> {
+        let tx_undo = tx
+            .outputs()
+            .iter()
+            .filter_map(|output| match output.purpose() {
+                OutputPurpose::StakePool(data) => Some((data, output.value())),
+                OutputPurpose::Transfer(_)
+                | OutputPurpose::LockThenTransfer(_, _)
+                | OutputPurpose::Burn => None,
+            })
+            .map(
+                |(pool_data, output_value)| -> Result<PoSAccountingUndo, ConnectTransactionError> {
+                    let tx_id = tx.get_id();
+                    let total_inputs = self.calculate_total_inputs(tx.inputs())?;
+                    ensure!(
+                        total_inputs.contains_key(&CoinOrTokenId::Coin),
+                        ConnectTransactionError::MissingCoinOutputToStake
+                    );
+                    ensure!(
+                        total_inputs.len() == 1,
+                        ConnectTransactionError::TokenInputForPoSAccountingOperation(tx_id)
+                    );
+                    // TODO: check StakePoolData fields
+
+                    let input0 = tx.inputs().get(0).expect("must be some");
```

Comment:
Please avoid using expect in such short logic situations, because as demonstrated in the call, removing the upper part, which we'll remove, will basically break the invariant.